File: src/hyperhandler/models/validator.py
"""Signal validation logic."""

from dataclasses import dataclass, field
from decimal import Decimal

from hyperhandler.models.signal import TradingSignal


@dataclass
class ValidationConfig:
    """Configuration for signal validation."""

    max_position_size_usd: Decimal = Decimal("10000")
    max_leverage: int = 20
    min_order_size: Decimal = Decimal("0.0001")
    require_stop_loss: bool = False
    allowed_pairs: list[str] = field(default_factory=list)  # Empty = all allowed


@dataclass
class ValidationResult:
    """Result of signal validation."""

    valid: bool
    errors: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)


class SignalValidator:
    """Validates trading signals against configured limits."""

    def __init__(self, config: ValidationConfig | None = None):
        self.config = config or ValidationConfig()

    def validate(
        self,
        signal: TradingSignal,
        current_price: Decimal | None = None,
    ) -> ValidationResult:
        """Validate a trading signal.

        Args:
            signal: The trading signal to validate.
            current_price: Current market price for size validation.

        Returns:
            ValidationResult with errors and warnings.
        """
        errors: list[str] = []
        warnings: list[str] = []

        # Check allowed pairs
        if self.config.allowed_pairs:
            if signal.pair not in self.config.allowed_pairs:
                errors.append(
                    f"Pair {signal.pair} not in allowed list: {self.config.allowed_pairs}"
                )

        # Check leverage
        if signal.leverage > self.config.max_leverage:
            errors.append(
                f"Leverage {signal.leverage} exceeds maximum {self.config.max_leverage}"
            )

        # Check minimum order size
        if signal.size < self.config.min_order_size:
            errors.append(
                f"Size {signal.size} below minimum {self.config.min_order_size}"
            )

        # Check position size in USD (if price available)
        price = current_price or signal.entry_price
        if price is not None:
            position_usd = signal.size * price
            if position_usd > self.config.max_position_size_usd:
                errors.append(
                    f"Position size ${position_usd:.2f} exceeds maximum "
                    f"${self.config.max_position_size_usd}"
                )

        # Check stop-loss requirement
        if self.config.require_stop_loss and signal.stop_loss is None:
            errors.append("Stop-loss is required but not provided")

        # Warnings for risky setups
        if signal.leverage > 10:
            warnings.append(f"High leverage ({signal.leverage}x) - increased risk")

        if signal.stop_loss is None:
            warnings.append("No stop-loss set - position has unlimited downside risk")

        return ValidationResult(
            valid=len(errors) == 0,
            errors=errors,
            warnings=warnings,
        )

    def validate_or_raise(
        self,
        signal: TradingSignal,
        current_price: Decimal | None = None,
    ) -> None:
        """Validate signal and raise exception if invalid.

        Args:
            signal: The trading signal to validate.
            current_price: Current market price for size validation.

        Raises:
            ValueError: If validation fails.
        """
        result = self.validate(signal, current_price)
        if not result.valid:
            raise ValueError(f"Signal validation failed: {'; '.join(result.errors)}")

File: internal/models/validator.go
package models

import (
	"fmt"
	"strings"

	"github.com/shopspring/decimal"

	_ "github.com/skilus/hyperhandler/internal/decimalx" // set DivisionPrecision=28
)

// ValidationConfig bounds what signals the validator accepts. The zero value is
// not the default — use NewValidationConfig (or DefaultValidationConfig) for the
// Python defaults. Mirrors validator.py:ValidationConfig.
type ValidationConfig struct {
	MaxPositionSizeUSD decimal.Decimal
	MaxLeverage        int
	MinOrderSize       decimal.Decimal
	RequireStopLoss    bool
	AllowedPairs       []string // empty = all allowed
}

// DefaultValidationConfig returns the Python default limits.
func DefaultValidationConfig() ValidationConfig {
	return ValidationConfig{
		MaxPositionSizeUSD: decimal.RequireFromString("10000"),
		MaxLeverage:        20,
		MinOrderSize:       decimal.RequireFromString("0.0001"),
		RequireStopLoss:    false,
	}
}

// ValidationResult is the outcome of validation. Mirrors
// validator.py:ValidationResult.
type ValidationResult struct {
	Valid    bool
	Errors   []string
	Warnings []string
}

// SignalValidator checks signals against configured limits. Mirrors
// validator.py:SignalValidator.
type SignalValidator struct {
	Config ValidationConfig
}

// NewSignalValidator returns a validator with the given config. Pass a nil
// pointer for the default config.
func NewSignalValidator(cfg *ValidationConfig) *SignalValidator {
	if cfg == nil {
		c := DefaultValidationConfig()
		return &SignalValidator{Config: c}
	}
	return &SignalValidator{Config: *cfg}
}

// Validate checks the signal and returns the collected errors and warnings.
// currentPrice (nil to omit) overrides the entry price for USD-size checks. The
// rule order mirrors validator.py:validate.
func (v *SignalValidator) Validate(signal *TradingSignal, currentPrice *decimal.Decimal) ValidationResult {
	var errs, warns []string

	// Allowed pairs.
	if len(v.Config.AllowedPairs) > 0 && !contains(v.Config.AllowedPairs, signal.Pair) {
		errs = append(errs, fmt.Sprintf("Pair %s not in allowed list: %s",
			signal.Pair, formatPairList(v.Config.AllowedPairs)))
	}

	// Leverage.
	if signal.Leverage > v.Config.MaxLeverage {
		errs = append(errs, fmt.Sprintf("Leverage %d exceeds maximum %d",
			signal.Leverage, v.Config.MaxLeverage))
	}

	// Minimum order size.
	if signal.Size.Cmp(v.Config.MinOrderSize) < 0 {
		errs = append(errs, fmt.Sprintf("Size %s below minimum %s",
			signal.Size.String(), v.Config.MinOrderSize.String()))
	}

	// Position size in USD, if a price is available.
	price := currentPrice
	if price == nil {
		price = signal.EntryPrice
	}
	if price != nil {
		positionUSD := signal.Size.Mul(*price)
		if positionUSD.Cmp(v.Config.MaxPositionSizeUSD) > 0 {
			errs = append(errs, fmt.Sprintf("Position size $%s exceeds maximum $%s",
				positionUSD.StringFixed(2), v.Config.MaxPositionSizeUSD.String()))
		}
	}

	// Stop-loss requirement.
	if v.Config.RequireStopLoss && signal.StopLoss == nil {
		errs = append(errs, "Stop-loss is required but not provided")
	}

	// Warnings for risky setups.
	if signal.Leverage > 10 {
		warns = append(warns, fmt.Sprintf("High leverage (%dx) - increased risk", signal.Leverage))
	}
	if signal.StopLoss == nil {
		warns = append(warns, "No stop-loss set - position has unlimited downside risk")
	}

	return ValidationResult{
		Valid:    len(errs) == 0,
		Errors:   errs,
		Warnings: warns,
	}
}

// ValidateOrRaise returns an error if validation fails, mirroring
// validator.py:validate_or_raise.
func (v *SignalValidator) ValidateOrRaise(signal *TradingSignal, currentPrice *decimal.Decimal) error {
	result := v.Validate(signal, currentPrice)
	if !result.Valid {
		return fmt.Errorf("Signal validation failed: %s", strings.Join(result.Errors, "; "))
	}
	return nil
}

func contains(xs []string, s string) bool {
	for _, x := range xs {
		if x == s {
			return true
		}
	}
	return false
}

// formatPairList renders the allowed pairs like Python's list repr, e.g.
// ['BTC', 'ETH'].
func formatPairList(pairs []string) string {
	quoted := make([]string, len(pairs))
	for i, p := range pairs {
		quoted[i] = "'" + p + "'"
	}
	return "[" + strings.Join(quoted, ", ") + "]"
}
